### PR TITLE
Support applyOnReload for LoadConfiguration

### PR DESCRIPTION
**Pull request by 304NotModified**
Fixes https://github.com/NLog/NLog/issues/1429

Todo 

- [ ] Fix default
- [ ] Event should be before loading
- [ ] Check other PR for event

### CODE REVIEW

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 1]
```diff
@@ -272,7 +272,10 @@ public override IEnumerable<string> FileNamesToWatch
         public override LoggingConfiguration Reload()
         {
             if (!string.IsNullOrEmpty(_originalFileName))
-                return new XmlLoggingConfiguration(_originalFileName, LogFactory);
+            {
+                LogFactory.LoadConfiguration(_originalFileName);
```
Comment (304NotModified) [inline src/NLog/Config/XmlLoggingConfiguration.cs:276]: FYI, needed this otherwise IAppEnvironment will be ignored

File: src/NLog/SetupBuilderExtensions.cs [hunk 2]
```diff
@@ -107,6 +115,16 @@ public static ISetupBuilder LoadConfiguration(this ISetupBuilder setupBuilder, A
                 setupBuilder.LogFactory.Configuration = newConfig;
             }
 
+            if (applyOnReload)
+            {
+#if !NETSTANDARD1_3
+                setupBuilder.LogFactory.ConfigurationChanged += (sender, e) =>
```
Comment (304NotModified) [inline src/NLog/SetupBuilderExtensions.cs:121]: I'm always a bit confused by Changed and Reloaded :'(

Seems Reloaded only works for file reloads?

Not sure if I need a test for that case (file reload) then

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/Config/XmlLoggingConfiguration.cs:276]: Breaking change as the Reload now assigns the Configuration. What happens if the LogFactory.LoadConfiguration fails?

File: src/NLog/SetupBuilderExtensions.cs (hunk 2 above)
Comment (snakefoot) [reply src/NLog/SetupBuilderExtensions.cs:121]: How do you ensure that you don't have multiple callbacks? Think users will expect that previous delegates will be reset, if calling LoadConfiguration twice.

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs:276]: > Breaking change as the Reload now assigns the Configuration

Well I always was under the impression that was a bug. Or do you think it was useful to have that separated? 

> No longer possible to call explicit Reload() ,and then adjust the returned config and then assign the config.

Why is that not possible now?

Comment (snakefoot) [reply src/NLog/Config/XmlLoggingConfiguration.cs:276]: Because `LogFactory.LoadConfiguration` will assign the configuration, so any adjustments are too late.

Comment (snakefoot) [reply src/NLog/Config/XmlLoggingConfiguration.cs:276]: But I can certainly agree that the config-reload is full of pitfalls. But not sure if the fix should be implemented on XmlLoggingConfiguration directly.

File: src/NLog/SetupBuilderExtensions.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/SetupBuilderExtensions.cs:121]: Could you give an example?
```
  logFactory.Setup().
                LoadConfigurationFromFile().
                LoadConfiguration(true, b =>
                {
                    b.Configuration.AddRuleForAllLevels("target1");
                })
                .LoadConfiguration(true, b =>
                {
                    b.Configuration.AddRuleForAllLevels("target2");
                });
```

Comment (snakefoot) [reply src/NLog/SetupBuilderExtensions.cs:121]: I was just thinking it might confuse if doing Setup()-operation more than once. But lets see what happens. Btw. can you make sure that the callback is only called when a new config is loaded? (Avoid callback with null on shutdown / configuration assigned to null).

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs:276]: > But maybe the improvement can be implemented without changing XmlLoggingConfiguration.Reload() ?

Good idea, will look into it. Will be tomorrow I guess

File: src/NLog/SetupBuilderExtensions.cs (hunk 2 above)
Comment (304NotModified) [reply src/NLog/SetupBuilderExtensions.cs:121]: > might confuse if doing Setup()-operation more than once

No this is great! Always wanted this! OK, maybe it needs a reset, but you could do that in LoadConfiguration

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs]: Fixed, please double check 

PS I would appreciate an Approve  "flag"  :)

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 3]
```diff
@@ -272,7 +272,7 @@ public override IEnumerable<string> FileNamesToWatch
         public override LoggingConfiguration Reload()
         {
             if (!string.IsNullOrEmpty(_originalFileName))
-                return new XmlLoggingConfiguration(_originalFileName, LogFactory);
+                return LogFactory.CreateConfig(_originalFileName); //TODO NLog 5: should call LoadConfiguration which set LogFactory.Configuration
```
Comment (snakefoot) [inline src/NLog/Config/XmlLoggingConfiguration.cs:275]: Still curious why you need to modify this method? Why not just have a method called `LogFactory.ReloadConfiguration()` instead of the config assigning itself?

File: src/NLog/SetupBuilderExtensions.cs [hunk 4]
```diff
@@ -106,6 +114,15 @@ public static ISetupBuilder LoadConfiguration(this ISetupBuilder setupBuilder, A
             {
                 setupBuilder.LogFactory.Configuration = newConfig;
             }
+#if !NETSTANDARD1_3
+            if (applyOnReload)
+            {
+                setupBuilder.LogFactory.ConfigurationChanged += (sender, e) =>
```
Comment (snakefoot) [inline src/NLog/SetupBuilderExtensions.cs:120]: ConfigurationChanged will give notifications on the following events:
- AutoReload - FileWatcher-logic
- ManualLoad - Assigning `LogFactory.Configuration = newConfig;`
- ManualUnload - Assigning `LogFactory.Configuration = null;`
- Shutdown

I think you should not perform the callback when there is no active LoggingConfiguration available. And maybe consider if the callback should also happen on `ManualLoad`.

File: tests/NLog.UnitTests/Config/LogFactorySetupTests.cs [hunk 5]
```diff
@@ -568,5 +569,103 @@ public void SetupBuilderLoadConfigurationFromXmlPatchTest()
             Assert.False(logFactory.AutoShutdown);
             Assert.Single(logFactory.Configuration.Variables);
         }
+
```
Comment (snakefoot) [inline tests/NLog.UnitTests/Config/LogFactorySetupTests.cs:572]: Maybe include a test-method where you call LogFactory.Close() ? (Similar to Shutdown).

**Comment by snakefoot** [comment]
Could also be interesting to verify that this works together with the autoreload-feature in MEL NLogLoggingConfiguration:

Notice how I abuse ConfigurationChanged to unhook from filewatch-notications from MEL-config-system:

https://github.com/NLog/NLog.Extensions.Logging/blob/master/src/NLog.Extensions.Logging/Config/NLogLoggingConfiguration.cs

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 3 above)
Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: because `new XmlLoggingConfiguration(_originalFileName, LogFactory);` doens't use the _configLoader, and thus you well get this.

```c#
   var appEnv = new Mocks.AppEnvironmentMock( ...)
  var configLoader = new LoggingConfigurationFileLoader(appEnv);
  var logFactory = new LogFactory(configLoader);

  // Act
  logFactory.Setup().
  LoadConfigurationFromFile().
  // config loaded from configLoader 

  // crash:  System.IO.FileNotFoundException - Could not find file 'D:\nlog\NLog\tests\NLog.UnitTests\bin\Debug\net452\NLog.config'.
  logFactory.Configuration = logFactory.Configuration.Reload()
```

File: src/NLog/SetupBuilderExtensions.cs (hunk 4 above)
Comment (304NotModified) [reply src/NLog/SetupBuilderExtensions.cs:120]: I'm also doubting with ManualLoad , but that one is hard to prevent now isn't?

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: Would it not be better to change `XmlLoggingConfiguration` to use the LogFactory-AppEnvironment for loading the XmlConfig-file ? (Improve mocking all around, instead of just Reload)

File: src/NLog/SetupBuilderExtensions.cs (hunk 4 above)
Comment (snakefoot) [reply src/NLog/SetupBuilderExtensions.cs:120]: > with ManualLoad , but that one is hard to prevent now isn't?

I guess for XmlLoggingConfiguration then one could check if config was assigned using LogFactory.Configuration-property-setter (And signal that through the ConfigurationChanged-EventArgs). But for NLogLoggingConfiguration then it is not easy, since it has to use the official setter-gateway.

So yes hard to distinguish AutoReload and ManulLoad.

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 3 above)
Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: Good point

File: src/NLog/SetupBuilderExtensions.cs (hunk 4 above)
Comment (304NotModified) [reply src/NLog/SetupBuilderExtensions.cs:120]: Good to know is that KeepVariablesOnReload is also working for manual load.

I have some things to fix here, but we should have a clear direction on this. If manual load is a no-go, I can't fix this (or not sure how)

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 1 above)
Comment (snakefoot) [reply src/NLog/Config/XmlLoggingConfiguration.cs]: Have created #3941 that improves the mocking of `XmlReader.Create` for `XmlLoggingConfiguration`.

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 3 above)
Comment (snakefoot) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: Have created #3941 that improves the mocking of `XmlReader.Create` for `XmlLoggingConfiguration`.

Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs]: > Why not in the future just have a method called LogFactory.ReloadConfiguration() instead of the config assigning itself?

I think also this is better! I think using the setter was a mistake, also I really need to use the reload event, not the changed. So fixed that

**Comment by 304NotModified** [comment]
@snakefoot please your opinion on this approach.

I still have to do:

- [ ] verify that this works together with the autoreload-feature in MEL
- [ ] change LoggingConfiguration.Reload()? Dunno how, revert doesn't work yet.
- [ ]  LogFactory.Close() / LogFactory.Dispose() ? (Similar to Shutdown) - this isn't needed anymore?

File: src/NLog/Config/XmlLoggingConfiguration.cs [hunk 6]
```diff
@@ -272,7 +272,7 @@ public override IEnumerable<string> FileNamesToWatch
         public override LoggingConfiguration Reload()
         {
             if (!string.IsNullOrEmpty(_originalFileName))
-                return new XmlLoggingConfiguration(_originalFileName, LogFactory);
+                return LogFactory.CreateConfig(_originalFileName);
```
Comment (snakefoot) [inline src/NLog/Config/XmlLoggingConfiguration.cs:275]: Still curious why this change is needed?

File: src/NLog/LogFactory.cs [hunk 7]
```diff
@@ -1166,6 +1168,20 @@ public LogFactory LoadConfiguration(string configFile)
...
         }
 
+        /// <summary>
+        /// Reload and apply the current config
+        /// </summary>
+        /// <returns>LogFactory instance for fluent interface</returns>
+        public LogFactory ReloadConfiguration()
+        {
+            InternalLogger.Info("Reloading configuration...");
+            Configuration = Configuration.ReloadNewConfig();
```
Comment (snakefoot) [inline src/NLog/LogFactory.cs:1178]: Could this feature be made in a separate PullRequest?

File: src/NLog/LogManager.cs [hunk 8]
```diff
@@ -172,6 +172,15 @@ public static LogFactory LoadConfiguration(string configFile)
             return factory;
         }
 
+        /// <summary>
+        /// Reload and apply the current config
+        /// </summary>
+        /// <returns>LogFactory instance for fluent interface</returns>
+        public static LogFactory ReloadConfiguration()
```
Comment (snakefoot) [inline src/NLog/LogManager.cs:179]: Could this feature be made in a separate PullRequest?

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 6 above)
Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: Because the test still fails, reload will search on the disk and fails

File: src/NLog/LogFactory.cs (hunk 7 above)
Comment (304NotModified) [reply src/NLog/LogFactory.cs:1178]: Will do

File: src/NLog/Config/XmlLoggingConfiguration.cs (hunk 6 above)
Comment (snakefoot) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: Even when you merge/rebase with master, and take in #3941 ?

Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: yes I did a rebase.

Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: See https://github.com/NLog/NLog/compare/logFactory.ReloadConfiguration...master

Comment (snakefoot) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: Really strange that it tries to access the disk when it can only access the disk using the mocked AppEnvironment. Are you mocking AppEnvironment ?

Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: I think the mocking is correct.

See branch loadconfiguration-applyOnReload_noLogFactory.CreateConfig

image

Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs:275]: I see the issue, LogFactory needs IAppEnvironment, while it's already in ILoggingConfigurationLoader

Comment (304NotModified) [reply src/NLog/Config/XmlLoggingConfiguration.cs]: Fixed

File: src/NLog/SetupBuilderExtensions.cs [hunk 9]
```diff
@@ -91,6 +91,14 @@ public static ISetupBuilder SetupSerialization(this ISetupBuilder setupBuilder,
...
         /// </summary>
         public static ISetupBuilder LoadConfiguration(this ISetupBuilder setupBuilder, Action<ISetupLoadConfigurationBuilder> configBuilder)
+        {
+            return LoadConfiguration(setupBuilder, true, configBuilder);
+        }
+
+        /// <summary>
+        /// Loads NLog config created by the method <paramref name="configBuilder"/>
+        /// </summary>
+        public static ISetupBuilder LoadConfiguration(this ISetupBuilder setupBuilder, bool applyOnReload, Action<ISetupLoadConfigurationBuilder> configBuilder)
```
Comment (304NotModified) [inline src/NLog/SetupBuilderExtensions.cs:101]: I'm doubting if adding a bool here is a good idea. Maybe it should be an option object? Otherwise it's hard to extend. 

What do you think @snakefoot ?

Comment (snakefoot) [reply src/NLog/SetupBuilderExtensions.cs:101]: When creating the LoadConfiguration-methods here https://github.com/NLog/NLog.Web/pull/540, then I wanted them to look like `AddJsonFile`. https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.configuration.jsonconfigurationextensions.addjsonfile But the lambda makes it a little different. If you believe the number of options for the same method will be more than 3 then I think an options-object makes sense.

File: src/NLog/LogFactory.cs [hunk 10]
```diff
@@ -1196,6 +1212,12 @@ internal LogFactory LoadConfiguration(string configFile, bool optional)
             return this;
         }
 
+        internal LoggingConfiguration CreateConfig(string configFile)
```
Comment (snakefoot) [inline src/NLog/LogFactory.cs:1215]: Method extraction is no loading needed, and I like calling `_configLoader.Load` better than the name `CreateConfig`.

File: src/NLog/SetupBuilderExtensions.cs [hunk 11]
```diff
@@ -91,6 +91,14 @@ public static ISetupBuilder SetupSerialization(this ISetupBuilder setupBuilder,
         /// Loads NLog config created by the method <paramref name="configBuilder"/>
         /// </summary>
         public static ISetupBuilder LoadConfiguration(this ISetupBuilder setupBuilder, Action<ISetupLoadConfigurationBuilder> configBuilder)
+        {
+            return LoadConfiguration(setupBuilder, true, configBuilder);
```
Comment (snakefoot) [inline src/NLog/SetupBuilderExtensions.cs:95]: Don't think it should be `true` by default.

File: src/NLog/SetupBuilderExtensions.cs [hunk 12]
```diff
@@ -106,6 +114,13 @@ public static ISetupBuilder LoadConfiguration(this ISetupBuilder setupBuilder, A
             {
                 setupBuilder.LogFactory.Configuration = newConfig;
             }
+            if (applyOnReload)
+            {
+                setupBuilder.LogFactory.ConfigurationReloaded += (sender, e) =>
+                {
+                    setupBuilder.LoadConfiguration(false, configBuilder);
```
Comment (snakefoot) [inline src/NLog/SetupBuilderExtensions.cs:121]: Should only perform the call when `e.Succeeded == true`.

Comment (snakefoot) [reply src/NLog/SetupBuilderExtensions.cs:121]: When using `ConfigurationReloaded` instead of `ConfigurationChanged`, then it will not work together with MEL-NLogLoggingConfiguration. As it doesn't make use of internal NLog file-watchers https://github.com/NLog/NLog.Extensions.Logging/blob/4bd3bcb9aceca3da3b76dee6cbe711f0cd0fb34b/src/NLog.Extensions.Logging/Config/NLogLoggingConfiguration.cs#L111

Think you should change from `ConfigurationReloaded` to `ConfigurationChanged`. Maybe mark `ConfigurationReloaded` as obsolete in NLog 5.0?

Comment (snakefoot) [reply src/NLog/SetupBuilderExtensions.cs:121]: @304NotModified Created #3954 with example of how to use `ConfigurationChanged`

**Comment by 304NotModified** [comment]
inactive so close for now
